code/datums/abilities/wraith.dm

```diff
@@ -147,44 +147,83 @@
 		return 0
 
-
 	doCooldown()         //This makes it so wraith early game is much faster but hits a wall of high absorb cooldowns after ~5 corpses
 		if (!holder)	 //so wraiths don't hit scientific notation rates of regen without playing perfectly for a million years
-			return
+			return		 //Edit: corpsecount is now added in absorb_corpse. - Josephin
 		var/datum/abilityHolder/wraith/W = holder
 		if (istype(W))
 			if (W.corpsecount == 0)
 				cooldown = 45 SECONDS
-				W.corpsecount += 1
 			else
 				cooldown += W.corpsecount * 150
-				W.corpsecount += 1
 		last_cast = world.time + cooldown
 		holder.updateButtons()
 		SPAWN(cooldown + 5)
 			holder.updateButtons()
 
+/datum/action/bar/icon/absorb_corpse //Yes this will be private, this is just for testing purposes.
+	duration = 10 SECONDS
+	interrupt_flags = INTERRUPT_ACT | INTERRUPT_ACTION | INTERRUPT_MOVE
+	id = "wraith_absorbCorpse"
+	icon = 'icons/mob/mob.dmi'
+	icon_state = "wraith"
+	var/mob/living/carbon/human/M
+	var/datum/targetable/wraithAbility/absorbCorpse/absorb
+	var/turf/wraith_loc
+
+	New(U, Absorb)
+		M = U
+		absorb = Absorb
+		..()
+
+
+	onUpdate()
+		..()
+		var/datum/abilityHolder/W = absorb.holder
+		var/datum/abilityHolder/wraith = absorb.holder.owner
+		if(W == null || M == null || wraith_loc != get_turf(absorb.holder.owner.loc))
+			interrupt(INTERRUPT_ALWAYS)
+			boutput(wraith, __red("Your attempt to draw essence from the corpse was interrupted!"))
+			return
+
+	onStart()
+		..()
+		var/datum/abilityHolder/W = absorb.holder
+		var/datum/abilityHolder/wraith = absorb.holder.owner
+		if(W == null || M == null)
+			interrupt(INTERRUPT_ALWAYS)
+			boutput(wraith, __red("Your attempt to draw essence from the corpse was interrupted!"))
+			return
+		boutput(wraith, __red("You start sucking the essence out of [M]'s corpse!"))
+		particleMaster.SpawnSystem(new /datum/particleSystem/localSmoke("#000000", 12, locate(M.x, M.y, M.z)))
+		for (var/mob/living/V in viewers(7, W.owner))
+			boutput(V, "Black smoke rises from [M]'s corpse! Freaky!")
+
+
+	onEnd()
+		..()
+		var/datum/abilityHolder/wraithHolder = absorb.holder
+		logTheThing("combat", wraithHolder, null, "absorbs the corpse of [key_name(M)] as a wraith.")
+		M.decomp_stage = 4
+		if (M.organHolder && M.organHolder.brain)
+			qdel(M.organHolder.brain)
+		M.set_face_icon_dirty()
+		M.set_body_icon_dirty()
+
+		wraithHolder.regenRate *= 2.0
+		wraithHolder.owner:onAbsorb(M)
+		//Messages for everyone!
+		boutput(wraithHolder.owner, "<span class='alert'><strong>[pick("You draw the essence of death out of [M]'s corpse!", "You drain the last scraps of life out of [M]'s corpse!")]</strong></span>")
+		playsound(M, "sound/voice/wraith/wraithsoulsucc[rand(1, 2)].ogg", 60, 0)
+		for (var/mob/living/V in viewers(7, wraithHolder.owner))
+			boutput(V, "[M]'s corpse suddenly rots to nothing but bone!")
+		var/datum/abilityHolder/wraith/wraithCorpse = absorb.holder
+		wraithCorpse.corpsecount += 1
+		return
```

Comment:
```suggestion
```

This doesn't do anything since it's a blank return at the end of a proc- the proc does that anyways